File: 4. 寻找两个正序数组的中位数.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        nums1 = self.mergeSort(nums1, nums2)
        print(nums1)
        length = len(nums1)
        return nums1[length >> 1] if (length & 1) else (nums1[length >> 1] + nums1[(length >> 1) - 1]) / 2.0
        
    def mergeSort(self, left, right):
        leftLen = len(left)
        rightLen = len(right)
        leftIndex = 0
        rightIndex = 0
        ret = []
        
        while((leftLen > leftIndex) and (rightLen > rightIndex)):
            if (left[leftIndex] < right[rightIndex]):
                ret.append(left[leftIndex])
                leftIndex += 1
            else:
                ret.append(right[rightIndex])
                rightIndex += 1
        
        if (leftLen > leftIndex):
            ret = ret + left[leftIndex:]
        if (rightLen > rightIndex):
            ret = ret + right[rightIndex:]
            
        return ret
    
x = Solution()
print(x.findMedianSortedArrays([1, 2], [3, 4]))

File: 4.寻找两个正序数组的中位数.js
/*
 * @lc app=leetcode.cn id=4 lang=javascript
 *
 * [4] 寻找两个正序数组的中位数
 */

// @lc code=start
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function (nums1, nums2) {
    nums1 = mergeSort(nums1, nums2);
    const len = nums1.length;
    return len % 2 === 0 ? (nums1[len / 2 - 1] + nums1[len / 2]) / 2 : nums1[Math.floor(len / 2)];
};

// 解法 1. 先归并排序，在找出来
function mergeSort(leftArr, rightArr) {
    let leftArrLen = leftArr.length,
        rightArrLen = rightArr.length,
        leftIndex = 0,
        rightIndex = 0,
        ret = [];

    while (leftArrLen > leftIndex && rightArrLen > rightIndex) {
        if (leftArr[leftIndex] <= rightArr[rightIndex]) {
            ret.push(leftArr[leftIndex++]);
        } else {
            ret.push(rightArr[rightIndex++]);
        }
    }

    if (leftIndex < leftArrLen) {
        ret.push(...leftArr.slice(leftIndex));
    }
    if (rightIndex < rightArrLen) {
        ret.push(...rightArr.slice(rightIndex));
    }
    return ret;
}
// @lc code=end

console.log(findMedianSortedArrays([2], []));
